src/test/java/io/cryostat/recordings/RecordingArchiveHelperTest.java

```diff
@@ -492,6 +492,59 @@ public Object answer(InvocationOnMock invocation) throws Throwable {
         Mockito.verify(notification).send();
     }
 
+    @Test
+    void shouldNotSaveEmptyRecording() throws Exception {
+
+        Mockito.when(connection.getService()).thenReturn(service);
+        IRecordingDescriptor descriptor = Mockito.mock(IRecordingDescriptor.class);
+        Mockito.when(descriptor.getName()).thenReturn(recordingName);
+
+        ServiceRef serviceRef1 =
+                new ServiceRef(
+                        URIUtil.convert(
+                                new JMXServiceURL(
+                                        "service:jmx:rmi:///jndi/rmi://cryostat:9091/jmxrmi")),
+                        "some.Alias.1");
+        ServiceRef serviceRef2 =
+                new ServiceRef(
+                        URIUtil.convert(
+                                new JMXServiceURL(
+                                        "service:jmx:rmi:///jndi/rmi://cryostat:9092/jmxrmi")),
+                        "some.Alias.2");
+        ServiceRef serviceRef3 =
+                new ServiceRef(
+                        URIUtil.convert(
+                                new JMXServiceURL(
+                                        "service:jmx:rmi:///jndi/rmi://cryostat:9093/jmxrmi")),
+                        "some.Alias.3");
+
+        Mockito.when(platformClient.listDiscoverableServices())
+                .thenReturn(List.of(serviceRef1, serviceRef2, serviceRef3));
+        Mockito.when(connection.getJMXURL())
+                .thenReturn(
+                        (new JMXServiceURL("service:jmx:rmi:///jndi/rmi://cryostat:9092/jmxrmi")));
+        Mockito.when(base32.encodeAsString(Mockito.any())).thenReturn("encodedServiceUri");
+
+        Instant now = Instant.now();
+        Mockito.when(clock.now()).thenReturn(now);
+        Mockito.when(fs.exists(Mockito.any())).thenReturn(true).thenReturn(false);
+        InputStream stream = Mockito.mock(InputStream.class);
+        Mockito.when(service.openStream(descriptor, false)).thenReturn(stream);
+        Path specificRecordingsPath = Mockito.mock(Path.class);
+        Mockito.when(archivedRecordingsPath.resolve(Mockito.anyString()))
+                .thenReturn(specificRecordingsPath);
+        Path destination = Mockito.mock(Path.class);
+        Mockito.when(specificRecordingsPath.resolve(Mockito.anyString())).thenReturn(destination);
+
+        Mockito.when(fs.copy(stream, specificRecordingsPath)).thenThrow(new IOException());
```

Comment:
After rebasing, the unit test `shouldNoSaveRecording()` in `RecordingArchiverHelperTest` is failing with the following message. I tried adding `throws IOException` and `throws EmptyRecordingException` to the method signature `writeRecordingToDestination` without noticing a change.

The failure also refers to line 115, which is during the `setup()` for notificationBuilder. How is this related to line 539?
```
[ERROR] shouldNotSaveEmptyRecording  Time elapsed: 0.023 s  <<< ERROR!
org.mockito.exceptions.base.MockitoException: 

Checked exception is invalid for this method!
Invalid: io.cryostat.recordings.EmptyRecordingException: java.io.IOException
	at io.cryostat.net.web.http.api.v1.TargetRecordingPatchSaveTest.shouldNotSaveEmptyRecording(TargetRecordingPatchSaveTest.java:115)
```

Reply:
Line 115 of that class refers to:

```java
        Mockito.when(recordingArchiveHelper.saveRecording(Mockito.any(), Mockito.any()))
>               .thenThrow(new EmptyRecordingException(new IOException()));
```

The problem is that after Hareet's last PR, the method signature of `recordingArchiveHelper.saveRecording` has changed, and it no longer throws any checked exceptions. Instead, it returns a `Future`, and any exception that occurs during the call would be propagated to the caller by `future.completeExceptionally`, so the caller would see an unchecked `ExecutionException` thrown.

Instead of the mock throwing an `EmptyRecordingException` directly it should return a `CompletableFuture.failedFuture`. This will fix the test error and put the handler down the expected execution path.

Reply:
The other test failure I see on your branch looks like this:

```
[ERROR]   RecordingArchiveHelperTest.shouldNotSaveEmptyRecording:541 Unexpected exception type thrown ==> expected: <io.cryostat.recordings.EmptyRecordingException> but was: <org.mockito.exceptions.misusing.PotentialStubbingProblem>
```

and further up the test logs, there is:

```
[ERROR] shouldNotSaveEmptyRecording  Time elapsed: 0.037 s  <<< FAILURE!
org.opentest4j.AssertionFailedError: Unexpected exception type thrown ==> expected: <io.cryostat.recordings.EmptyRecordingException> but was: <org.mockito.exceptions.misusing.PotentialStubbingProblem>
	at org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:65)
	at org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:37)
	at org.junit.jupiter.api.Assertions.assertThrows(Assertions.java:3007)
	at io.cryostat.recordings.RecordingArchiveHelperTest.shouldNotSaveEmptyRecording(RecordingArchiveHelperTest.java:541)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:688)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$6(TestMethodTestDescriptor.java:210)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:206)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:131)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:65)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:143)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:143)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:220)
	at org.junit.platform.launcher.core.DefaultLauncher.lambda$execute$6(DefaultLauncher.java:188)
	at org.junit.platform.launcher.core.DefaultLauncher.withInterceptedStreams(DefaultLauncher.java:202)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:181)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:128)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:150)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:124)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
Caused by: org.mockito.exceptions.misusing.PotentialStubbingProblem: 
Strict stubbing argument mismatch. Please check:
 - this invocation of 'copy' method:
    fs.copy(
    Mock for InputStream, hashCode: 2023197227,
    Mock for Path, hashCode: 949506330
);
    -> at io.cryostat.recordings.RecordingArchiveHelper.writeRecordingToDestination(RecordingArchiveHelper.java:343)
 - has following stubbing(s) with different arguments:
    1. fs.copy(
    Mock for InputStream, hashCode: 2023197227,
    Mock for Path, hashCode: 902890505
);
      -> at io.cryostat.recordings.RecordingArchiveHelperTest.shouldNotSaveEmptyRecording(RecordingArchiveHelperTest.java:539)
Typically, stubbing argument mismatch indicates user mistake when writing tests.
Mockito fails early so that you can debug potential problem easily.
However, there are legit scenarios when this exception generates false negative signal:
  - stubbing the same method multiple times using 'given().will()' or 'when().then()' API
    Please use 'will().given()' or 'doReturn().when()' API for stubbing.
  - stubbed method is intentionally invoked with different arguments by code under test
    Please use default or 'silent' JUnit Rule (equivalent of Strictness.LENIENT).
For more information see javadoc for PotentialStubbingProblem class.
	at io.cryostat.core.sys.FileSystem.copy(FileSystem.java:91)
	at io.cryostat.recordings.RecordingArchiveHelper.writeRecordingToDestination(RecordingArchiveHelper.java:343)
	at io.cryostat.recordings.RecordingArchiveHelperTest.lambda$shouldNotSaveEmptyRecording$2(RecordingArchiveHelperTest.java:543)
	at org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:55)
	... 68 more
```

Looking at the surrounding context around line `543`, this looks related to the exception message:


```java
539:        Mockito.when(fs.copy(stream, specificRecordingsPath)).thenThrow(new IOException());
```

It seems that `fs.copy(stream, specificRecordingsPath))` is causing the mock stubbing error, because `specificRecordingsPath` is used to configure the stub behaviour, but when the test runs the mock's `copy` method is called with a different second argument than the `specificRecordingsPath` variable. Most likely there is some difference now between how this test expects `fs.copy` to be called and what path it is provided vs. what the handler is actually doing. This makes sense, since Hareet's last PR that you're rebasing onto changes how these paths are handled in a pretty significant way.